teps/0036-start-measuring-tekton-pipelines-performance.md

```diff
@@ -0,0 +1,99 @@
+last-updated: '2020-11-20'
+authors:
+- '@bobcatfish'
+---
+
+# TEP-0036: Start Measuring Tekton Pipelines Performance
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Up until this point, we have left it to our users to report back to us on how well Tekton Pipelines performs. Relying on
+our users in this way means that we can’t easily inform users about what performance to expect and also means that we
+usually only catch and fix performance issues after releases.
+
+This proposal is all about finding an incremental step forward we can take towards being more proactive in responding to
+performance issues, and will also enable us to be able to evaluate the performance impact of implementation decisions
+going forward.
+
+Issue: [tektoncd/pipeline#540](https://github.com/tektoncd/pipeline/issues/540)
+
+## Motivation
+
+* To be able to understand and communicate about Tekton performance on an ongoing basis
+* Be able to answer questions like:
+  * How much overhead does a Task add?
+  * How much overhead does a Pipeline add?
+  * How much overhead does using a Tekton Bundle vs referencing a Task in my cluster add?
+  * If we switch out X component for Y component in the controller implementation, how does this impact performance?
+  * How much better or worse does release X perform than release Y?
+  * What characteristics should I look for when choosing what kind of machine to tun the Tekton Pipelines controller on?
+ * Being able to get ahead of issues such as [tektoncd/pipeline#3521](https://github.com/tektoncd/pipeline/issues/3521)
+
+### Goals
+
+The goal of this proposal is to avoid trying to boil the performance and loadtesting ocean all at once, and to instead
+identify some (relatively) small step we can take towards starting to gather this kind of information, so we can build
+from there.
+
+* Identify Service Level Indicators for Tekton Pipelines (as few as possible to start with, maybe just one)
+* For each Service Level Indicator define a target range (SLO) for some known setup
+* Setup infrastructure required such that:
+  * Contributors and users can find the data they need on Tekton Performance
+  * Performance is measured regularly at some interval (e.g. daily, weekly, per release)
+  * Reliable and actionable alerting is setup (if feasible) to notify maintainers when SLOs are violated
+
+Reference: [Definitions of SLIs and SLOs](https://landing.google.com/sre/sre-book/chapters/service-level-objectives/).
+These are worded in terms of running observable services. Since Tekton Pipelines is providing a service that can be run
+by others (vs hosting a "Tekton Pipelines" instance we expect users to use) our situation is a bit different, but I
+think the same SLI and SLO concepts can be applied.
```

Comment:
I agree that the concepts can be applied but I wonder if we should just call them Performance Indicators and Performance Objectives?  Reason being that SLI + SLO implies (to my ear) SLA.  And SLA is kinda more of a contractual thing that sometimes get used by corporations to bash each other over the head or hold back payments or get money back or w/e.

What we're offering here isn't really contractual I don't think?  So I wonder if just naming them something slightly different (while 100% still linking to the SRE book as reference material for meaning) will null some of that impression?

Conversely, let's say we _do_ stick with the SLI/SLO language.  What's the SLA?  Or, in other words, what are the outcomes for the Tekton Pipelines project when a performance regression is identified?  Do we hold back releases?  Drop all other work to focus on it?  Have a dedicated build captain role specific to performance?  Perhaps this also needs to be a Goal / Requirement that we have to hash out in WGs?

Reply:
yup I agree, I like Performance Indicators and Objectives better, great suggestion @sbwsg.

Reply:
I'm tempted to stick with SLO and SLI b/c they are well known terms, and switching to performance indicators and objectives feels a bit like we're trying to invent something new when we don't have to.

@sbwsg I think you're totally right that SLO and SLI are often used in the same context as SLA but really I don't think they imply SLA

Reply:
sgtm!

Reply:
sure, I am open going with SLO and SLI, we can always change it if needed.